src/mtcp/mtcp_restart.c

```diff
@@ -496,13 +502,92 @@ mysetauxval(char **evp, unsigned long int type, unsigned long int val)
   return -1;
 }
 
-#define PAGESIZE 4096 /* We hardwire this, since we can't make libc calls. */
-// We choose this address based on our observation on CORI that it won't overlap
-// with any other memory segments. We might change this address in the future,
-// for now, it doesn't matters.
-#define vvarStartTmp ((0x40000) - 5*PAGESIZE)
-#define vdsoStartTmp ((0x40000) - 2*PAGESIZE)
+/* We hardwire these, since we can't make libc calls. */
+#define PAGESIZE 4096
+#define ROUNDADDRUP(addr, size) ((addr + size - 1) & ~(size - 1))
+// I think we're forced to use global variables, since
+// unmap_memory_areas_and_restore_vdso uses these variables subsequently.
+uint64_t vvarStartTmp = 0;
+uint64_t vdsoStartTmp = 0;
+
+// This function searches for a free memory region to temporarily remap the vdso
+// and vvar regions to, so that mtcp's vdso and vvar do not overlap with the
+// checkpointed process' vdso and vvar.
+static void
+remap_vdso_and_vvar_regions() {
+  Area area;
+  int rc = 0;
+  uint64_t vvarStart = 0;
+  uint64_t vdsoStart = 0;
+  uint64_t vvarSize = 0;
+  uint64_t vdsoSize = 0;
+  uint64_t prev_addr = 0x10000;
+
+  int mapsfd = mtcp_sys_open2("/proc/self/maps", O_RDONLY);
+  if (mapsfd < 0) {
+    MTCP_PRINTF("error opening /proc/self/maps; errno: %d\n", mtcp_sys_errno);
+    mtcp_abort();
+  }
+
+  while (mtcp_readmapsline(mapsfd, &area)) {
+    if (mtcp_strcmp(area.name, "[vvar]") == 0) {
+      vvarStart = area.addr;
+      vvarSize = area.size;
+    }
+
+    if (mtcp_strcmp(area.name, "[vdso]") == 0) {
+      vdsoStart = area.addr;
+      vdsoSize = area.size;
+    }
 
+    if (vvarStartTmp == 0) {
+      // In May 2020, on Cori and elsewhere, vvar is 3 pages and vdso is 2
+      // pages.
+      if (prev_addr + 5 * PAGESIZE <= area.addr) {
+        vvarStartTmp = prev_addr;
+        vdsoStartTmp = prev_addr + 3 * PAGESIZE;
+      } else {
+        prev_addr = ROUNDADDRUP((uint64_t) area.endAddr, PAGESIZE);
+      }
+    }
+
+    if (vvarStart > 0 && vdsoStart > 0 && vvarStartTmp > 0 && vdsoStartTmp > 0)
+    {
+      break;
+    }
+  }
+
+  mtcp_sys_close(mapsfd);
+
+  if ((vvarStart > 0 && vvarStartTmp == 0) ||
+    (vdsoStartTmp > 0 && vdsoStartTmp == 0)) {
+    MTCP_PRINTF("No free region found to temporarily map vvar/vdso to\n");
+    mtcp_abort();
+  }
+
+  if (vvarStart > 0) {
+    rc = mtcp_sys_mremap(vvarStart, vvarSize, vvarSize,
+                         MREMAP_FIXED | MREMAP_MAYMOVE, vvarStartTmp);
+  }
+
+  if (rc == MAP_FAILED) {
```

Comment:
Usually, we want to check it right after we assign it.

When `vvarStart == 0`, it makes no sense to check it, and easy to break when someone else need to add something in this function.